sympy/solvers/solvers.py

```diff
@@ -2305,60 +2306,150 @@ def solve_linear_system(system, *symbols, **flags):
+
     if isinstance(equ, Eq):
-        # got equation, so move all the
-        # terms to the left hand side
-        equ = equ.lhs - equ.rhs
+        eq = equ.lhs - equ.rhs
+    else:
+        eq = equ
+
+    ceq = cancel(eq)
+    xeq = _mexpand(ceq.as_numer_denom()[0], recursive=True)
+
+    free = xeq.free_symbols
+    coeffs = free & set(coeffs)
+    if not coeffs:
+        return []  # solve(0, x) -> []
+
+    if not syms:
+        # e.g. A*exp(x) + B - (exp(x) + y) separated into parts that
+        # don't/do depend on coeffs gives
+        # -(exp(x) + y), A*exp(x) + B
+        # then see what symbols are common to both
+        # {x} = {x, A, B} - {x, y}
+        ind, dep = xeq.as_independent(*coeffs, as_Add=True)
+        dfree = dep.free_symbols
+        syms = dfree & ind.free_symbols
+        if not syms:
+            # but if the system looks like (a + b)*x + b - c
+            # then {} = {a, b, x} - c
+            # so calculate {x} = {a, b, x} - {a, b}
+            syms = dfree - set(coeffs)
+        if not syms:
+            syms = [Dummy()]
+        sym_dict = {}
+    else:
+        if len(syms) == 1 and iterable(syms[0]):
+            syms = syms[0]
+        e, s, sym_dict = recast_to_symbols([xeq], syms)
+        xeq = e[0]
+        syms = s
 
-    equ = cancel(equ).as_numer_denom()[0]
+    # find the functional forms in which symbols appear
 
-    system = list(collect(equ.expand(), sym, evaluate=False).values())
+    gens = set(xeq.as_coefficients_dict(*syms).keys()) - {1}
 
-    if not any(equ.has(sym) for equ in system):
-        # consecutive powers in the input expressions have
-        # been successfully collected, so solve remaining
-        # system using Gaussian elimination algorithm
-        _flags = dict(flags, set=None, dict=True)
-        sol = solve(system, *coeffs, **_flags)
-        if len(sol) == 1:
-            return sol[0]
-        elif not sol:
-            return []
+    # make sure we are working with symbols for generators
+
+    e, s, d = recast_to_symbols([xeq], list(gens))
+    xeq = e[0]
+    gens = s
+
+    # collect coefficients in front of generators
+
+    system = list(collect(xeq, gens, evaluate=False).values())
+
+    # get a solution
+
+    _flags = dict(flags, set=None, dict=True)
+    soln = solve(system, coeffs, **_flags)
+
+    # simplify because this flag is ignored by solve for a linear system
+
+    if flags.get('simplify', True):
+        soln = [{sym_dict.get(k, k): v.simplify()
+            for k, v in s.items()} for s in soln]
+
+    # unpack as necessary
+    coeffs = list(ordered(coeffs))
+    if not soln:
+        return [] if not flags.get('set', False) else (coeffs, [])
+    if flags.get('dict', False):
+        return soln
+    if flags.get('set', False):
+        return (coeffs, [
+            tuple([s.get(i, i) for i in coeffs]) for s in soln])
+    if not soln:
+        return {}
+    elif len(soln) > 1:
+        return soln
+    else:
+        return soln[0]
```

Comment:
Is there a reason this can't just let solve handle the flags?

If it's for the simplify code above then is there actually a need to call `simplify` (given that `solve` already doesn't do it for linear systems)?

Reply:
Yes. If you request `simplify` you should get simplification. `solve` does not do so for poly systems (linear or non-linear) per #23889.

Reply:
Why should simplify be used here if not in solve?

Reply:
I don't like the idea of having flags whose action is not well defined. If you pass `simplify=True` you have no idea if the solution was simplified or not. So why even allow that flag? I'm ok with removing it. I don't think that having a function as a front end to solve is a good idea. There should be a function to extract the coefficient system of equations and then the user can pass this with desired flags to solve. *This* function should be deprecated.